odk/odk.py

```diff
@@ -528,6 +534,20 @@ def fill_missing(self):
         if len(released_products) > 0:
             self.released_products = released_products
 
+        for product in self.products:
+            if product.maintenance == "merged":
+                if product.source_mappings is None:
+                    # Merge all other non-merge sets to make this one
+                    product.source_mappings = [p.id for p in self.products if p.maintenance != "merged"]
+                else:
+                    # Check that all listed source sets exist
+                    for source in product.source_mappings:
+                        if source not in [p.id for p in self.products]:
+                            raise Exception(f"Unknown source mapping set '{source}'")
+            elif product.maintenance == "extract":
+                if product.source_file is None:
+                    product.source_file = "$(EDIT_PREPROCESSED)"
```

Comment:
I am assuming you thought this through - I recently struggled hard with a pipeline that provided a merged sssom mapping set which was partially extracted from upstream, but partially generated by the pipeline itself.. So, something like: build a bridge, then extract a mapping set from that, than merge that into a mapping set you publish. This caused some circularity. I don't think this is the case here, just wanted to share my experience.

Reply:
I am not sure I understand your concern, and how it applies here.

In any case, the only change here is that the logic to handle the absence of a `source_file` key has been moved from within the Makefile template to the `odk.py` script. This is so the template can simply says

```Jinja2
$(MAPPINGDIR)/{{ mapping.id }}.sssom.tsv: {{ mapping.source_file }}
```

instead of

```Jinja2
$(MAPPINGDIR)/{{ mapping.id }}.sssom.tsv: {% if mapping.source_file is not none %}{{ mapping.source_file }}{% else %}$(EDIT_PREPROCESSED){% endif %}
```

This is functionally identical to what we had before, only with a (slightly) more readable template.